p2p/pex/addrbook_test.go

```diff
@@ -36,6 +37,264 @@ func deleteTempFile(fname string) {
+	book.SetLogger(log.TestingLogger())
+	assert.Zero(t, book.Size())
+
+	N := 10
+	randAddrs := randNetAddressPairs(t, N)
+	// mark all addrs but one as good
+	for _, addr := range randAddrs[:N-1] {
+		book.AddAddress(addr.addr, addr.src)
+		book.MarkGood(addr.addr)
+	}
+	// add the last addr, don't mark as good
+	for _, addr := range randAddrs[N-1:] {
+		book.AddAddress(addr.addr, addr.src)
+	}
+
+	// ensure we can get selection.
+	// this used to result in an infinite loop (!)
+	addrs := book.GetSelectionWithBias(biasToSelectNewPeers)
+	assert.NotNil(t, addrs, "expected an address")
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func listsAreEqual(a, b []int) bool {
+	if (a == nil) != (b == nil) {
+		return false
+	}
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// Creates a Book with m old and n new addresses.
+func createAddrBook(t *testing.T, m, n int) *addrBook {
+	fname := createTempFileName("addrbook_test")
+	defer deleteTempFile(fname)
+
+	book := NewAddrBook(fname, true)
+	book.SetLogger(log.TestingLogger())
+	assert.Zero(t, book.Size())
+
+	// add m old addresses
+	randAddrs := randNetAddressPairs(t, m)
+	for _, addr := range randAddrs {
+		book.AddAddress(addr.addr, addr.src)
+		book.MarkGood(addr.addr)
+	}
+	// add n new addresses
+	randAddrs = randNetAddressPairs(t, n)
+	for _, addr := range randAddrs {
+		book.AddAddress(addr.addr, addr.src)
+	}
+
+	return book
+}
+
+// Computes n * bias/100
+func numExpectedAddresses(bias, n int) int {
+	if n == 0 {
+		return 0
+	}
+	return int(math.Round((float64(bias) / float64(100)) * float64(n)))
+}
+
+// Counts the number of Old and New addresses
+func countOldAndNewInAddrs(addrs []*p2p.NetAddress, book *addrBook) (nold, nnew int) {
+	nold = 0
+	nnew = 0
+
+	for _, addr := range addrs {
+		if addr == nil {
+			continue
+		}
+		if book.IsGood(addr) {
+			nold++
+		} else {
+			nnew++
+		}
+	}
+	return nold, nnew
+}
+
+func analyseSelectionLayout(book *addrBook, addrs []*p2p.NetAddress) ([]int, []int, error) {
+	// address types are: 0 - nil, 1 - new, 2 - old
+	prevt := 0
+	var seqLens []int
+	var seqTypes []int
+	cnt := 0
+	for i, addr := range addrs {
+		if addr == nil {
+			err := fmt.Errorf("nil addresses in selection, index %d", i)
+			return nil, nil, err
+		}
+		addrt := 0
+		if book.IsGood(addr) {
+			addrt = 2
+		} else {
+			addrt = 1
+		}
+
+		if addrt != prevt && prevt != 0 {
+			seqLens = append(seqLens, cnt)
+			seqTypes = append(seqTypes, prevt)
+			cnt = 0
+		}
+		cnt++
+		prevt = addrt
+	}
+	seqLens = append(seqLens, cnt)
+	seqTypes = append(seqTypes, prevt)
+	return seqLens, seqTypes, nil
+}
+
+func TestEmptyAddrBookSelection(t *testing.T) {
+	book := createAddrBook(t, 0, 0)
+	addrs := book.GetSelectionWithBias(biasToSelectNewPeers)
+	assert.Nil(t, addrs, "expected a nill selection")
+}
+
+func testAddrBookAddressSelection(t *testing.T, b int) {
```

Comment:
this test (as of now) is a bit complicated and will be hard to modify in the future

Reply:
I looked at how to make it better but short of cutting out from the checks, couldn't find a way. I am opened to suggestions.